File: src/rsmm/sdk/health.py
"""Health system — boot canary + crash history, as written by the LOADER.

The loader is the only process that can observe a crashy boot, so it owns the
file. `src/loader/src/health.cpp` opens `<game>/mods/_health.json` before any
mod code runs, stamps `canary.step` as each `init.lua` executes, and closes the
canary ~2 s after `ready`. A canary still open at the next launch means the
previous run died at that step, so the crash is attributed to that mod; three
consecutive failed boots disable it.

    {
      "version": 1,
      "canary": {"open": false, "step": "boot_ok", "session": "fd1d"},
      "mods": {
        "Foo": {"crashes": 2, "last_error": "...",
                "disabled": false, "disabled_reason": ""}
      }
    }

This module is the read/write side for the CLI. It used to describe — and
address — a completely different pair of files: `.rsmm_boot.json` and
`.rsmm_health.json` under `_Cooking`, with `last_step` instead of
`canary.step` and `disabled_by_health` instead of `disabled`. Nothing has ever
written those, so every consumer silently no-opped: `rsmm doctor` reported "no
crash records" while the loader had a mod disabled, `apply`'s quarantine pass
never quarantined anything, and `safe-mode --bisect` wrote its findings where
the loader would never read them. Two health systems that never met.

Writes are read-modify-write on the whole document so the CLI can never clobber
the canary node the loader is keeping, and land via temp-file + rename — the
whole point of this file is to survive a process that dies at an arbitrary
instant, so it must never be observed truncated.
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass, field
from pathlib import Path

from .api import sdk_export

#: Written by the loader, next to the mods it describes.
HEALTH_FILE_NAME = "_health.json"

#: Mirrors `kStrikeLimit` in src/loader/src/health.cpp. The loader does the
#: disabling; this is here so the CLI can explain the rule and so
#: `record_crash` (safe-mode bisect) reaches the same verdict.
DEFAULT_THRESHOLD = 3


@dataclass
class ModHealth:
    crashes: int = 0
    last_error: str = ""
    disabled: bool = False
    disabled_reason: str = ""
    last_seen: int = 0


@dataclass
class HealthState:
    threshold: int = DEFAULT_THRESHOLD
    mods: dict[str, ModHealth] = field(default_factory=dict)


def _resolve_game_dir(path: Path) -> Path:
    """Accept the game dir or the `_Cooking` dir; return the game dir.

    Every existing caller passes `<game>/DarkTalesResources/_Cooking`, because
    that is where this class used to keep its own files. Rather than touch each
    one, normalise here.
    """
    path = Path(path)
    if (path / "Ravenswatch.exe").exists() or (path / "mods").is_dir():
        return path
    # <game>/DarkTalesResources/_Cooking -> <game>
    if path.name == "_Cooking" and len(path.parents) >= 2:
        return path.parents[1]
    return path


class Health:
    """Read/write the loader's crash history and inspect its boot canary."""

    def __init__(self, path: Path):
        self.game_dir = _resolve_game_dir(path)
        self.mods_dir = self.game_dir / "mods"
        self.health_path = self.mods_dir / HEALTH_FILE_NAME
        #: Kept for callers that still speak in terms of the cooking dir.
        self.cooking = Path(path)

    # ---- raw document --------------------------------------------------

    def _read_doc(self) -> dict:
        try:
            raw = json.loads(self.health_path.read_text(encoding="utf-8"))
        except (OSError, ValueError):
            return {}
        return raw if isinstance(raw, dict) else {}

    def _write_doc(self, doc: dict) -> None:
        self.mods_dir.mkdir(parents=True, exist_ok=True)
        tmp = self.health_path.with_suffix(self.health_path.suffix + ".tmp")
        tmp.write_text(json.dumps(doc, indent=2) + "\n", encoding="utf-8")
        os.replace(tmp, self.health_path)

    # ---- health sidecar -----------------------------------------------

    def load(self) -> HealthState:
        doc = self._read_doc()
        st = HealthState()
        mods = doc.get("mods")
        if not isinstance(mods, dict):
            return st
        for mid, body in mods.items():
            if not isinstance(body, dict):
                continue
            st.mods[str(mid)] = ModHealth(
                crashes=int(body.get("crashes", 0) or 0),
                last_error=str(body.get("last_error", "") or ""),
                disabled=bool(body.get("disabled", False)),
                disabled_reason=str(body.get("disabled_reason", "") or ""),
                last_seen=int(body.get("last_seen", 0) or 0),
            )
        return st

    def save(self, st: HealthState) -> None:
        # Read-modify-write: `canary` and `version` belong to the loader and
        # must survive a CLI write. Rebuilding the document from HealthState
        # alone would drop the canary mid-session and lose the attribution for
        # a crash that had already happened.
        #
        # This makes a CLI write safe against the file on disk, which is only
        # half the race: the loader keeps the whole document in memory for the
        # session, so it used to overwrite a CLI edit landing mid-launch (the
        # desktop app running `safe-mode --reset <id>` with the game open saw
        # the mod re-quarantine itself). health.cpp::save_locked now merges
        # onto the current file and only writes back the quarantine fields for
        # mods THAT session quarantined, so both directions hold.
        doc = self._read_doc()
        doc["version"] = doc.get("version", 1)
        doc["mods"] = {
            mid: {
                "crashes": h.crashes,
                "last_error": h.last_error,
                "disabled": h.disabled,
                "disabled_reason": h.disabled_reason,
                **({"last_seen": h.last_seen} if h.last_seen else {}),
            }
            for mid, h in sorted(st.mods.items())
        }
        self._write_doc(doc)

    # ---- boot canary --------------------------------------------------

    def read_canary(self) -> dict | None:
        """Return the canary node if one is OPEN, else None.

        An open canary is itself the crash signal: the loader closes it once a
        launch has demonstrably survived boot. A closed one is not interesting,
        so it reads the same as no canary at all.
        """
        canary = self._read_doc().get("canary")
        if not isinstance(canary, dict) or not canary.get("open", False):
            return None
        return canary

    def clear_canary(self) -> None:
        doc = self._read_doc()
        canary = doc.get("canary")
        if not isinstance(canary, dict):
            return
        canary["open"] = False
        doc["canary"] = canary
        self._write_doc(doc)

    def attribute_crash(self, canary: dict) -> str | None:
        """Given an open canary, return the mod id we hold responsible."""
        step = str(canary.get("step", ""))
        if step.startswith("per_mod:"):
            return step[len("per_mod:"):]
        return None  # crash before any mod ran -> not a mod's fault

    @sdk_export("Health.record_crash")
    def record_crash(self, mod_id: str, error: str = "") -> HealthState:
        """Bump the mod's crash counter, persist, return the updated state.

        At `threshold` the mod is marked disabled, which the loader honours at
        load and the applier honours at apply.
        """
        st = self.load()
        h = st.mods.setdefault(mod_id, ModHealth())
        h.crashes += 1
        h.last_error = error[:512]
        h.last_seen = int(time.time())
        if h.crashes >= st.threshold and not h.disabled:
            h.disabled = True
            h.disabled_reason = f"failed to boot {h.crashes} times in a row"
        self.save(st)
        return st

    @sdk_export("Health.disabled_mods")
    def disabled_mods(self) -> set[str]:
        """Ids of mods the loader (or a bisect) has quarantined.

        These were disabled after a crashy launch; the user re-enables each
        with :meth:`re_enable` once fixed.
        """
        return {mid for mid, h in self.load().mods.items() if h.disabled}

    @sdk_export("Health.re_enable")
    def re_enable(self, mod_id: str) -> None:
        """User manually re-enables after fixing the crash."""
        st = self.load()
        h = st.mods.get(mod_id)
        if not h:
            return
        h.crashes = 0
        h.disabled = False
        h.disabled_reason = ""
        h.last_error = ""
        self.save(st)

File: src/loader/src/health.cpp
#include "health.h"
#include "loader.h"

#include "json.hpp"

#include <cstring>
#include <fstream>
#include <set>
#include <string>
#include <mutex>

namespace fs = std::filesystem;

namespace rsmm {
namespace health {
namespace {

// Consecutive failed boots before a mod is skipped at load.
constexpr int kStrikeLimit = 3;

std::mutex g_mu;
bool       g_ready = false;
fs::path   g_path;
std::string g_session;
nlohmann::json g_doc;

// Mods whose quarantine state THIS session set. `disabled`/`disabled_reason`
// are the one part of the document both writers touch: the loader sets them on
// auto-quarantine, the CLI clears them (`safe-mode --reset`) and sets them
// (`--bisect`). g_doc is a session-long in-memory copy, so a plain rewrite
// silently reverted any CLI edit made while the game was running — the desktop
// app resetting a mod with the game open saw it re-quarantine itself. Only ids
// in here have their quarantine fields written back; everyone else's come from
// disk. See save_locked().
std::set<std::string> g_owned_quarantine;

nlohmann::json& mods_node() {
    if (!g_doc.contains("mods") || !g_doc["mods"].is_object()) {
        g_doc["mods"] = nlohmann::json::object();
    }
    return g_doc["mods"];
}

nlohmann::json& mod_node(const std::string& id) {
    auto& mods = mods_node();
    if (!mods.contains(id) || !mods[id].is_object()) {
        mods[id] = nlohmann::json{{"crashes", 0}, {"last_error", ""},
                                  {"disabled", false}, {"disabled_reason", ""}};
    }
    return mods[id];
}

// Re-read what is on disk right now, so a save merges onto the CURRENT
// document rather than onto whatever it looked like when this process started.
// Returns an empty object if the file is missing or unparsable (a torn file
// must not stop the loader from recording a crash).
nlohmann::json read_disk_doc() {
    std::error_code ec;
    if (g_path.empty() || !fs::exists(g_path, ec)) return nlohmann::json::object();
    nlohmann::json disk;
    try {
        std::ifstream f(g_path);
        f >> disk;
    } catch (const std::exception&) {
        return nlohmann::json::object();
    }
    return disk.is_object() ? disk : nlohmann::json::object();
}

// Temp-file + rename: the whole point of this file is to survive a process
// that dies at an arbitrary instant, so it must never be observed truncated.
//
// The write is a MERGE, not a rewrite. The loader owns `canary`, `crashes` and
// `last_error`; the CLI owns the quarantine fields except for the mods this
// session quarantined itself (g_owned_quarantine). Anything else on disk —
// including a mod this process never saw — is carried through untouched.
void save_locked() {
    if (g_path.empty()) return;

    nlohmann::json out = read_disk_doc();
    out["version"] = 1;
    if (g_doc.contains("canary")) out["canary"] = g_doc["canary"];

    if (!out.contains("mods") || !out["mods"].is_object()) {
        out["mods"] = nlohmann::json::object();
    }
    const auto mine = g_doc.find("mods");
    if (mine != g_doc.end() && mine->is_object()) {
        for (auto it = mine->begin(); it != mine->end(); ++it) {
            if (!it.value().is_object()) continue;
            auto& dst = out["mods"][it.key()];
            if (!dst.is_object()) dst = nlohmann::json::object();
            dst["crashes"]    = it.value().value("crashes", 0);
            dst["last_error"] = it.value().value("last_error", std::string{});
            const bool ours = g_owned_quarantine.count(it.key()) != 0;
            if (ours || !dst.contains("disabled")) {
                dst["disabled"] = it.value().value("disabled", false);
                dst["disabled_reason"] =
                    it.value().value("disabled_reason", std::string{});
            }
        }
    }

    const auto tmp = g_path.string() + ".tmp";
    {
        std::ofstream f(tmp, std::ios::trunc);
        if (!f) return;
        f << out.dump(2) << "\n";
        if (!f.good()) return;
    }
    std::error_code ec;
    fs::rename(tmp, g_path, ec);
    if (ec) fs::remove(tmp, ec);
}

void set_canary_locked(bool open, const std::string& step) {
    g_doc["canary"] = nlohmann::json{
        {"open", open}, {"step", step}, {"session", g_session}};
    save_locked();
}

} // namespace

void init(const fs::path& mods_dir, const std::string& session) {
    std::lock_guard<std::mutex> g(g_mu);
    if (g_ready) return;
    g_ready  = true;
    g_path   = mods_dir / "_health.json";
    g_session = session;

    std::error_code ec;
    if (fs::exists(g_path, ec)) {
        try {
            std::ifstream f(g_path);
            f >> g_doc;
        } catch (const std::exception& e) {
            Loader::get().log(std::string("[health] history unreadable, starting "
                                          "fresh: ") + e.what());
            g_doc = nlohmann::json::object();
        }
    }
    if (!g_doc.is_object()) g_doc = nlohmann::json::object();
    g_doc["version"] = 1;

    // An open canary means the previous launch never reached "boot survived".
    const auto canary = g_doc.value("canary", nlohmann::json::object());
    if (canary.is_object() && canary.value("open", false)) {
        const std::string step = canary.value("step", std::string("unknown"));
        const std::string prev_session = canary.value("session", std::string("?"));
        std::string culprit = "_loader";
        constexpr const char* kPrefix = "per_mod:";
        if (step.rfind(kPrefix, 0) == 0) culprit = step.substr(std::strlen(kPrefix));
        auto& node = mod_node(culprit);
        node["crashes"] = node.value("crashes", 0) + 1;
        node["last_error"] = "previous launch (session " + prev_session
                           + ") did not survive boot; last step was '" + step + "'";
        const int n = node.value("crashes", 0);
        Loader::get().log("[health] previous launch did not finish booting at step '"
                          + step + "' -> attributed to '" + culprit + "' (crash #"
                          + std::to_string(n) + ")");
        // Three consecutive failed boots is a mod that reliably bricks the
        // game — the user cannot reach any in-game UI to turn it off, so the
        // loader does it and says so loudly. Recoverable: clear the flag in
        // mods/_health.json (or re-enable from the manager) to try again.
        if (n >= kStrikeLimit && culprit != "_loader" && !node.value("disabled", false)) {
            node["disabled"] = true;
            node["disabled_reason"] =
                "failed to boot " + std::to_string(n) + " times in a row";
            g_owned_quarantine.insert(culprit);
            Loader::get().log("[health] '" + culprit + "' DISABLED after "
                              + std::to_string(n) + " failed boots; edit "
                              "mods/_health.json to re-enable");
        }
    }
    set_canary_locked(true, "boot");
}

void checkpoint(const std::string& step) {
    std::lock_guard<std::mutex> g(g_mu);
    if (!g_ready) return;
    set_canary_locked(true, step);
}

void mark_boot_ok() {
    std::lock_guard<std::mutex> g(g_mu);
    if (!g_ready) return;
    const auto canary = g_doc.value("canary", nlohmann::json::object());
    if (!canary.value("open", false)) return;   // already closed this session
    // The strike counter measures CONSECUTIVE failed boots, so a launch that
    // got through resets it — otherwise unrelated crashes months apart would
    // eventually add up to an auto-disable. last_error is kept for diagnosis.
    auto& mods = mods_node();
    for (auto it = mods.begin(); it != mods.end(); ++it) {
        if (it.value().is_object()) it.value()["crashes"] = 0;
    }
    set_canary_locked(false, "boot_ok");
    Loader::get().log("[health] boot canary closed (game survived load)");
}

namespace {
// Read-only lookup. `g_doc[key]` would INSERT a null on a missing key, which
// on a const-looking getter is both surprising and a silent file mutation.
const nlohmann::json* find_mod(const std::string& mod_id) {
    auto it = g_doc.find("mods");
    if (it == g_doc.end() || !it->is_object()) return nullptr;
    auto m = it->find(mod_id);
    if (m == it->end() || !m->is_object()) return nullptr;
    return &(*m);
}
} // namespace

int crash_count(const std::string& mod_id) {
    std::lock_guard<std::mutex> g(g_mu);
    if (!g_ready) return 0;
    const auto* node = find_mod(mod_id);
    return node ? node->value("crashes", 0) : 0;
}

std::string last_error(const std::string& mod_id) {
    std::lock_guard<std::mutex> g(g_mu);
    if (!g_ready) return {};
    const auto* node = find_mod(mod_id);
    return node ? node->value("last_error", std::string{}) : std::string{};
}

void note_error(const std::string& mod_id, const std::string& msg) {
    std::lock_guard<std::mutex> g(g_mu);
    if (!g_ready) return;
    mod_node(mod_id)["last_error"] = msg;
    save_locked();
}

void set_disabled(const std::string& mod_id, const std::string& reason) {
    std::lock_guard<std::mutex> g(g_mu);
    if (!g_ready) return;
    auto& node = mod_node(mod_id);
    node["disabled"] = true;
    node["disabled_reason"] = reason;
    g_owned_quarantine.insert(mod_id);
    save_locked();
    Loader::get().log("[health] '" + mod_id + "' marked disabled: " + reason);
}

bool is_disabled(const std::string& mod_id) {
    std::lock_guard<std::mutex> g(g_mu);
    if (!g_ready) return false;
    const auto* node = find_mod(mod_id);
    return node && node->value("disabled", false);
}

} // namespace health
} // namespace rsmm
